### ISSUE TITLE
core.remote_sudo prompts for a password when run

### ISSUE THREAD

**Issue by cbrenton** [StackStorm/st2#2899]
Even when a password is provided as a parameter, the `core.remote_sudo` action hangs after prompting the user for a password. The expected behavior would be the action being executed without prompting the user for a password.

**Comment by bigmstone** [comment]
I've reproduced this locally, and working to resolve it.

**Comment by bigmstone** [comment]
From my digging into the code it would appear the handling of `sudo` isn't very robust.

Stated in more detail we execute the command prepending `sudo -E -- bash -c` to it (https://github.com/StackStorm/st2/blob/master/st2common/st2common/models/system/paramiko_command_action.py#L46) but don't have any mechanism to catch if a password prompt is initiated. It would appear we assume the sudoer will have unprompted access to run commands.

The trail I followed is listed below:
[Runner gets command to run](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/remote_command_runner.py#L40v)
[Command object created](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/remote_command_runner.py#L63)
[Begin command execution](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/remote_command_runner.py#L43)
[Generate command](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/remote_command_runner.py#L50)
[Execute parallel ssh method with command](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/remote_command_runner.py#L51)
[Send `_run_command` method and options to `_execute_in_pool` ](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/ssh/parallel_ssh.py#L115)
[Spawn instance of `execute_method`](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/ssh/parallel_ssh.py#L216)
[Execute command](https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/ssh/parallel_ssh.py#L260)

I haven't found any logic to determine if password prompt is given for sudo, and if it is to put in the password. Perhaps this is by design and the st2 user is expected to provide a system user with no-password access to sudo.

**Comment by Kami** [comment]
@bigmstone @cbrenton correct, per documentation (and installation script) user needs to set up paswordless sudo for stanley user.

As far as "robustness" goes, I agree - we should make the code more robust and user-friendly. Previously, with the fabric runner we caught "sudo requires password" scenario and throw a more user-friendly exception which said Paswordless sudo needs to be set up for that user.

Right now I don't know exactly what happens, but I would imagine action times out while waiting on prompt...

**Comment by Kami** [comment]
Having said that, we could perhaps introduce support for sudo with password by adding a new "sudo_password" or similar parameter and updating the local and remote runner (or at least the remote one).

Docs for default installation would still require paswordless setup for user `stanley` though.

**Comment by bigmstone** [comment]
I can't easily conceive of a way to gracefully catch this that doesn't run the risk of blindly typing the user's password into the shell.

**Comment by bigmstone** [comment]
@cbrenton I've captured this as a feature request now. Will put it on the docket to add mechanism to handle both prompted and unprompted sudo.

**Comment by Kami** [comment]
Here is an example of how fabric handles sudo password (it also supports interactive mode and prompt which we don't need to support) - https://github.com/fabric/fabric/blob/master/fabric/io.py#L170

So yeah, not totally straight forward to implement in our remote runner... So something we probably don't have time to do right now, but user contributions are of course welcome :)
